README
Flask App Engine Blobstore

Flask extension module for working with the blobstore & files api on App Engine.


CHANGES
flask-gae_blobstore Changelog
=============================

v1.0.1, 2013-01-24 -- Fixed setup.py package_data.
v1.0.0, 2013-01-01 -- Initial release.


test/app.py
"""
  `dev_appserver.py .
  `curl -X POST -F 'file=@test.jpg' http://localhost:8080/test`
"""

from flask import json, request
from flask import Flask
import gae_blobstore
import logging

app = Flask(__name__)
app.debug = True

@app.route('/test', methods=gae_blobstore.OPTIONS)
@gae_blobstore.upload_blobs()
def test(blobs):
  logging.info(request.form)
  return json.dumps(blobs.to_dict())

if __name__ == '__main__':
  app.run()


test/gae_blobstore.py
../../lib/python2.7/site-packages/gae_blobstore.py

gae_blobstore_tests.py
#!/usr/bin/env python
try:
  # a hack to see if the app engine sdk is loaded..
  import yaml
except ImportError:
  import dev_appserver
  dev_appserver.fix_sys_path()

import unittest
import gae_blobstore
from flask import json
from flask import Flask
from flask.ext import gae_tests
from google.appengine.api import files
from google.appengine.ext import ndb


# test application..

class TestModel(ndb.Model):
  blob_key = ndb.BlobKeyProperty()

app = Flask(__name__)
app.debug = True
app.request_class = gae_tests.FileUploadRequest

@app.route('/test_upload', methods=['POST', 'OPTIONS', 'HEAD', 'PUT'])
@gae_blobstore.upload_blobs()
def test_upload(blobs):
  entities = []
  try:
    for blob in blobs:
      entity = TestModel(
        blob_key=blob.blob_key)
      entities.append(entity)
    ndb.put_multi(entities)
  except:
    # rollback the operation and delete the blobs,
    # so they are not orphaned..
    for blob in blobs:
      blob.delete()
    raise Exception('Saving file upload info to datastore failed..')
  return json.dumps(blobs.to_dict())


# test cases..

class TestCase(gae_tests.TestCase):

  def test_blobstore_sanity_check(self):
    filename = files.blobstore.create(mime_type='application/octet-stream')
    self.assertNotEquals(None, filename)
    with files.open(filename, 'a') as f:
      f.write('test blob data..')
    files.finalize(filename)
    blobkey = files.blobstore.get_blob_key(filename)
    self.assertNotEquals(None, blobkey)

  def test_upload_returns_valid_blob_result(self):
    data, filename, size = gae_tests.open_test_file('test.jpg')
    response = app.test_client().post(
      data={'test': (data, filename)},
      path='/test_upload',
      headers={},
      query_string={})
    self.assertEqual(200, response.status_code)
    result = json.loads(response.data)
    self.assertIsInstance(result, list)
    self.assertEquals(1, len(result))
    self.assertEquals(True, result[0].get('successful'))
    # check the file name is the same..
    self.assertEquals(filename, result[0].get('name'))
    # check file size is the same..
    self.assertEquals(size, result[0].get('size'))
    # validate the blob_key..
    self.assertTrue(len(result[0].get('blob_key')) > 0)


if __name__ == '__main__':
  unittest.main()


gae_upload.py
"""
  flaskext.gae_upload
  ~~~~~~~~~~~~~~~~~~~

  Flask extension module for working with file upload on App Engine.

  :copyright: (c) 2012 by gregorynicholas.
  :license: BSD, see LICENSE for more details.
"""
import re
import time
# Uses of a deprecated module 'string'
# pylint: disable-msg=W0402
import string
import random
import logging
from flask import Response, request
from werkzeug import exceptions
from functools import wraps
from google.appengine.api import files
from google.appengine.ext import blobstore

delete = blobstore.delete
delete_async = blobstore.delete_async
fetch_data = blobstore.fetch_data
fetch_data_async = blobstore.fetch_data_async

WRITE_MAX_RETRIES = 3
WRITE_SLEEP_SECONDS = 0.05
DEFAULT_NAME_LEN = 20
MSG_INVALID_FILE_POSTED = 'Invalid file posted.'

UPLOAD_MIN_FILE_SIZE = 1
UPLOAD_MAX_FILE_SIZE = 1024*1024
# set by default to images..
UPLOAD_ACCEPT_FILE_TYPES = re.compile('image/(gif|p?jpeg|jpg|(x-)?png|tiff)')

# todo: need a way to easily configure these values..
ORIGINS = '*'
OPTIONS = ['OPTIONS', 'HEAD', 'GET', 'POST', 'PUT']
HEADERS = ['Accept', 'Content-Type', 'Origin', 'X-Requested-With']
MIMETYPE = 'application/json'

class RemoteResponse(Response):
  '''Base class for remote service `Response` objects.'''
  default_mimetype = MIMETYPE
  def __init__(self, response=None, mimetype=None, *args, **kw):
    if mimetype is None:
      mimetype = self.default_mimetype
    Response.__init__(self, response=response, mimetype=mimetype, **kw)
    self._fixcors()

  def _fixcors(self):
    self.headers['Access-Control-Allow-Origin'] = ORIGINS
    self.headers['Access-Control-Allow-Methods'] = ', '.join(OPTIONS)
    self.headers['Access-Control-Allow-Headers'] = ', '.join(HEADERS)

class FieldResultSet(list):
  def to_dict(self):
    result = []
    for field in self:
      result.append(field.to_dict())
    return result

class FieldResult:
  def __init__(self, name, type, size, field, value):
    self.successful = False
    self.error_msg = ''
    self.blob_key = None
    self.name = name
    self.type = type
    self.size = size
    self.field = field
    self.value = value

  def to_dict(self):
    return {
      'successful': self.successful,
      'error_msg': self.error_msg,
      'blob_key': str(self.blob_key),
      'name': self.name,
      'type': self.type,
      'size': self.size,
      # these two are commented out so the class is easily json serializable..
      # 'field': self.field,
      # 'value': self.value,
    }

def upload_blobs(validations=None):
  '''Method decorator for writing posted files to the blobstore using the
  App Engine files api. Passes an argument to the method with a list of
  `FieldResult` with BlobKey, name, type, size for each posted input
  file.

    :param validations:
  '''
  def wrapper(func):
    @wraps(func)
    def decorated(*args, **kw):
      return func(blobs=save_blobs(
        fields=_upload_fields(), validations=validations), *args, **kw)
    return decorated
  return wrapper

def save_blobs(fields, validations=None):
  '''Returns a list of `FieldResult` with BlobKey, name, type, size for
  each posted file.

    :param fields:
  '''
  results = FieldResultSet()
  i = 0
  # validate_file_type=False,
  # validate_min_file_size=False,
  # validate_max_file_size=False
  for name, field in fields:
    value = field.stream.read()
    result = FieldResult(
      name=re.sub(r'^.*\\', '', field.filename),
      type=field.content_type,
      size=len(value),
      field=field,
      value=value)
    if not validate(result):
      result.successful = False
      result.error_msg = MSG_INVALID_FILE_POSTED
      logging.warn('Error in file upload: %s', result.error_msg)
    else:
      result.blob_key = write_to_blobstore(
        result.value, result.type, result.name)
      if result.blob_key:
        result.successful = True
      else:
        result.successful = False
      results.append(result)
    i += 1
  return results

def _upload_fields():
  '''Return a list of tuples with name of the file and stream as value.'''
  result = []
  for key, value in request.files.iteritems():
    if type(value) is not unicode:
      result.append((key, value))
  return result

def get_field_size(field):
  try:
    field.seek(0, 2) # Seek to the end of the file
    size = field.tell() # Get the position of EOF
    field.seek(0) # Reset the file position to the beginning
    return size
  except:
    return 0

def validate(field,
    accept_file_types=UPLOAD_ACCEPT_FILE_TYPES,
    min_file_size=UPLOAD_MIN_FILE_SIZE,
    max_file_size=UPLOAD_MAX_FILE_SIZE):
  '''Validates a file input based on size & type. If validation fails, adds
  an error property to the field arg dict.

    :param field:
    :param accept_file_types:
    :param min_file_size:
    :param max_file_size:
  '''
  if field.size < UPLOAD_MIN_FILE_SIZE:
    field.error_msg = 'min_file_size'
  elif field.size > UPLOAD_MAX_FILE_SIZE:
    field.error_msg = 'max_file_size'
  # only allow images to be posted to this handler
  elif not accept_file_types.match(field.type):
    field.error_msg = 'accept_file_types'
  else:
    return True
  return False

def write_to_blobstore(data, mime_type, name=None):
  '''Writes a file to the App Engine blobstore and returns an instance of a
  BlobKey if successful.

    :param data:
    :param mime_type:
    :param name:
  '''
  if not name:
    name = ''.join(random.choice(string.letters)
      for x in range(DEFAULT_NAME_LEN))
  try:
    blob = files.blobstore.create(
      mime_type=mime_type,
      _blobinfo_uploaded_filename=name)
    with files.open(blob, 'a', exclusive_lock=True) as f:
      f.write(data)
    files.finalize(blob)
    result = files.blobstore.get_blob_key(blob)
    # issue with the local development SDK. we can only write to the blobstore
    # so fast, so set a retry_count and delay the execution thread between
    # each attempt..
    for i in range(1, WRITE_MAX_RETRIES):
      if result:
        break
      else:
        logging.debug('blob still None.. will retry to write to blobstore..')
        time.sleep(WRITE_SLEEP_SECONDS)
        result = files.blobstore.get_blob_key(blob)
      logging.debug('File written to blobstore: key: "%s"', result)
    return result
  except:
    import traceback
    logging.exception('Exception writing to the blobstore: %s',
      traceback.format_exc())


searchindex.js
Search.setIndex({objects:{"":{flask_gae_blobstore:[0,0,1,""]},"flask_gae_blobstore.FieldResult":{to_dict:[0,3,1,""],"__init__":[0,3,1,""]},flask_gae_blobstore:{RemoteResponse:[0,2,1,""],FieldResultSet:[0,2,1,""],get_field_size:[0,1,1,""],save_blobs:[0,1,1,""],write_to_blobstore:[0,1,1,""],validate:[0,1,1,""],FieldResult:[0,2,1,""],upload_blobs:[0,1,1,""]},"flask_gae_blobstore.FieldResultSet":{to_dict:[0,3,1,""],"__init__":[0,3,1,""]},"flask_gae_blobstore.RemoteResponse":{"__init__":[0,3,1,""]}},terms:{upload_blob:0,accept_file_typ:0,help:0,xmpp:[],upload_accept_file_typ:[],remoterespons:0,paramet:0,member:[],flask:[0,1],param:0,add:0,gae_blobstor:[0,1],dict:0,"0x7ff9b5ae82c0":[],mail:[],pass:0,write_sleep_second:[],sourc:0,"return":0,save_blob:0,default_name_len:[],blobkei:0,mime_typ:0,fieldresultset:0,name:0,success:0,changelog:[0,1],list:0,to_dict:0,each:0,"0x7ff8db64b130":[],"0x7f995b15eb10":[],work:0,see:0,arg:0,fail:0,index:0,upload_min_file_s:[],blob_kei:0,method:0,signatur:0,gregorynichola:0,msg_invalid_file_post:[],extens:0,base:0,post:0,get_field_s:0,error_msg:0,valu:0,instanc:0,sre_pattern:0,origin:[],"_sre":0,app:0,header:[],api:0,min_file_s:0,size:0,upload_max_file_s:[],messag:[],master:0,write_max_retri:[],instal:0,option:[],"0x7fc01c979f10":[],"0x7fa11c071380":0,link:0,input:0,none:0,remot:0,type:0,servic:0,properti:0,gae_messag:[],write_to_blobstor:0,"0x7ff3f9a750f0":[],"0x7fac4d954260":[],error:0,fieldresult:0,max_file_s:0,respons:0,argument:0,pip:0,blobstor:0,file:0,tarbal:0,"__init__":0,"0x7fa1fe214b00":[],decor:0,engin:0,"0x7f89e30b3790":[],write:0,field:0,valid:0,document:0,http:0,flask_gae_blobstor:0,object:0,initi:0,data:0,"class":0,github:0,mimetyp:0,com:0},objtypes:{"0":"py:module","1":"py:function","2":"py:class","3":"py:method"},titles:["Flask App Engine Blobstore","flask-gae_blobstore Changelog"],objnames:{"0":["py","module","Python module"],"1":["py","function","Python function"],"2":["py","class","Python class"],"3":["py","method","Python method"]},filenames:["index","changelog"]})

gae_blobstore.py
"""
  gae_blobstore
  ~~~~~~~~~~~~~~~~~~~

  Flask extension module for working with the blobstore & files api on App Engine.

  :copyright: (c) 2012 by gregorynicholas.
  :license: BSD, see LICENSE for more details.
"""
import re
import time
# Uses of a deprecated module 'string'
# pylint: disable-msg=W0402
import string
import random
import logging
from flask import Response, request
from werkzeug import exceptions
from functools import wraps
from google.appengine.api import files
from google.appengine.ext import blobstore

delete = blobstore.delete
delete_async = blobstore.delete_async
fetch_data = blobstore.fetch_data
fetch_data_async = blobstore.fetch_data_async

WRITE_MAX_RETRIES = 3
WRITE_SLEEP_SECONDS = 0.05
DEFAULT_NAME_LEN = 20
MSG_INVALID_FILE_POSTED = 'Invalid file posted.'

UPLOAD_MIN_FILE_SIZE = 1
UPLOAD_MAX_FILE_SIZE = 1024*1024
# set by default to images..
UPLOAD_ACCEPT_FILE_TYPES = re.compile('image/(gif|p?jpeg|jpg|(x-)?png|tiff)')

# todo: need a way to easily configure these values..
ORIGINS = '*'
OPTIONS = ['OPTIONS', 'HEAD', 'GET', 'POST', 'PUT']
HEADERS = ['Accept', 'Content-Type', 'Origin', 'X-Requested-With']
MIMETYPE = 'application/json'

class RemoteResponse(Response):
  '''Base class for remote service `Response` objects.'''
  default_mimetype = MIMETYPE
  def __init__(self, response=None, mimetype=None, *args, **kw):
    if mimetype is None:
      mimetype = self.default_mimetype
    Response.__init__(self, response=response, mimetype=mimetype, **kw)
    self._fixcors()

  def _fixcors(self):
    self.headers['Access-Control-Allow-Origin'] = ORIGINS
    self.headers['Access-Control-Allow-Methods'] = ', '.join(OPTIONS)
    self.headers['Access-Control-Allow-Headers'] = ', '.join(HEADERS)

class FieldResultSet(list):
  def to_dict(self):
    result = []
    for field in self:
      result.append(field.to_dict())
    return result

class FieldResult:
  def __init__(self, name, type, size, field, value):
    self.successful = False
    self.error_msg = ''
    self.blob_key = None
    self.name = name
    self.type = type
    self.size = size
    self.field = field
    self.value = value

  def to_dict(self):
    return {
      'successful': self.successful,
      'error_msg': self.error_msg,
      'blob_key': str(self.blob_key),
      'name': self.name,
      'type': self.type,
      'size': self.size,
      # these two are commented out so the class is easily json serializable..
      # 'field': self.field,
      # 'value': self.value,
    }

def upload_blobs(validations=None):
  '''Method decorator for writing posted files to the blobstore using the
  App Engine files api. Passes an argument to the method with a list of
  `FieldResult` with BlobKey, name, type, size for each posted input
  file.

    :param validations:
  '''
  def wrapper(func):
    @wraps(func)
    def decorated(*args, **kw):
      return func(blobs=save_blobs(
        fields=_upload_fields(), validations=validations), *args, **kw)
    return decorated
  return wrapper

def save_blobs(fields, validations=None):
  '''Returns a list of `FieldResult` with BlobKey, name, type, size for
  each posted file.

    :param fields:
  '''
  results = FieldResultSet()
  i = 0
  # validate_file_type=False,
  # validate_min_file_size=False,
  # validate_max_file_size=False
  for name, field in fields:
    value = field.stream.read()
    result = FieldResult(
      name=re.sub(r'^.*\\', '', field.filename),
      type=field.content_type,
      size=len(value),
      field=field,
      value=value)
    if not validate(result):
      result.successful = False
      result.error_msg = MSG_INVALID_FILE_POSTED
      logging.warn('Error in file upload: %s', result.error_msg)
    else:
      result.blob_key = write_to_blobstore(
        result.value, result.type, result.name)
      if result.blob_key:
        result.successful = True
      else:
        result.successful = False
      results.append(result)
    i += 1
  return results

def _upload_fields():
  '''Return a list of tuples with name of the file and stream as value.'''
  result = []
  for key, value in request.files.iteritems():
    if type(value) is not unicode:
      result.append((key, value))
  return result

def get_field_size(field):
  try:
    field.seek(0, 2) # Seek to the end of the file
    size = field.tell() # Get the position of EOF
    field.seek(0) # Reset the file position to the beginning
    return size
  except:
    return 0

def validate(field,
    accept_file_types=UPLOAD_ACCEPT_FILE_TYPES,
    min_file_size=UPLOAD_MIN_FILE_SIZE,
    max_file_size=UPLOAD_MAX_FILE_SIZE):
  '''Validates a file input based on size & type. If validation fails, adds
  an error property to the field arg dict.

    :param field:
    :param accept_file_types:
    :param min_file_size:
    :param max_file_size:
  '''
  if field.size < UPLOAD_MIN_FILE_SIZE:
    field.error_msg = 'min_file_size'
  elif field.size > UPLOAD_MAX_FILE_SIZE:
    field.error_msg = 'max_file_size'
  # only allow images to be posted to this handler
  elif not accept_file_types.match(field.type):
    field.error_msg = 'accept_file_types'
  else:
    return True
  return False

def write_to_blobstore(data, mime_type, name=None):
  '''Writes a file to the App Engine blobstore and returns an instance of a
  BlobKey if successful.

    :param data:
    :param mime_type:
    :param name:
  '''
  if not name:
    name = ''.join(random.choice(string.letters)
      for x in range(DEFAULT_NAME_LEN))
  try:
    blob = files.blobstore.create(
      mime_type=mime_type,
      _blobinfo_uploaded_filename=name)
    with files.open(blob, 'a', exclusive_lock=True) as f:
      f.write(data)
    files.finalize(blob)
    result = files.blobstore.get_blob_key(blob)
    # issue with the local development SDK. we can only write to the blobstore
    # so fast, so set a retry_count and delay the execution thread between
    # each attempt..
    for i in range(1, WRITE_MAX_RETRIES):
      if result:
        break
      else:
        logging.debug('blob still None.. will retry to write to blobstore..')
        time.sleep(WRITE_SLEEP_SECONDS)
        result = files.blobstore.get_blob_key(blob)
      logging.debug('File written to blobstore: key: "%s"', result)
    return result
  except:
    import traceback
    logging.exception('Exception writing to the blobstore: %s',
      traceback.format_exc())


flask_gae_blobstore.py
"""
  flask_gae_blobstore
  ~~~~~~~~~~~~~~~~~~~

  Flask extension for working with the blobstore & files apis on
  App Engine.

  :copyright: (c) 2012 by gregorynicholas.
  :license: BSD, see LICENSE for more details.
"""
import re
import time
# Uses of a deprecated module 'string'
# pylint: disable-msg=W0402
import string
import random
import logging
from flask import Response, request
from werkzeug import exceptions
from functools import wraps
from google.appengine.api import files
from google.appengine.ext import blobstore

__all__ = ['delete', 'delete_async', 'fetch_data', 'fetch_data_async',
'WRITE_MAX_RETRIES', 'WRITE_SLEEP_SECONDS', 'DEFAULT_NAME_LEN',
'MSG_INVALID_FILE_POSTED', 'UPLOAD_MIN_FILE_SIZE', 'UPLOAD_MAX_FILE_SIZE',
'UPLOAD_ACCEPT_FILE_TYPES', 'ORIGINS', 'OPTIONS', 'HEADERS', 'MIMETYPE',
'RemoteResponse', 'FieldResultSet', 'FieldResult', 'upload_blobs',
'save_blobs', 'write_to_blobstore']

delete = blobstore.delete
delete_async = blobstore.delete_async
fetch_data = blobstore.fetch_data
fetch_data_async = blobstore.fetch_data_async

#:
WRITE_MAX_RETRIES = 3
#:
WRITE_SLEEP_SECONDS = 0.05
#:
DEFAULT_NAME_LEN = 20
#:
MSG_INVALID_FILE_POSTED = 'Invalid file posted.'

#:
UPLOAD_MIN_FILE_SIZE = 1
#:
UPLOAD_MAX_FILE_SIZE = 1024*1024
# set by default to images..
#:
UPLOAD_ACCEPT_FILE_TYPES = re.compile('image/(gif|p?jpeg|jpg|(x-)?png|tiff)')

# todo: need a way to easily configure these values..
#:
ORIGINS = '*'
#:
OPTIONS = ['OPTIONS', 'HEAD', 'GET', 'POST', 'PUT']
#:
HEADERS = ['Accept', 'Content-Type', 'Origin', 'X-Requested-With']
#:
MIMETYPE = 'application/json'

class RemoteResponse(Response):
  '''Base class for remote service `Response` objects.

    :param response:
    :param mimetype:
  '''
  default_mimetype = MIMETYPE
  def __init__(self, response=None, mimetype=None, *args, **kw):
    if mimetype is None:
      mimetype = self.default_mimetype
    Response.__init__(self, response=response, mimetype=mimetype, **kw)
    self._fixcors()

  def _fixcors(self):
    self.headers['Access-Control-Allow-Origin'] = ORIGINS
    self.headers['Access-Control-Allow-Methods'] = ', '.join(OPTIONS)
    self.headers['Access-Control-Allow-Headers'] = ', '.join(HEADERS)

class FieldResultSet(list):
  def to_dict(self):
    '''
      :returns:
    '''
    result = []
    for field in self:
      result.append(field.to_dict())
    return result

class FieldResult:
  '''
    :param successful:
    :param error_msg:
    :param blob_key:
    :param name:
    :param type:
    :param size:
    :param field:
    :param value:
  '''
  def __init__(self, name, type, size, field, value):
    self.successful = False
    self.error_msg = ''
    self.blob_key = None
    self.name = name
    self.type = type
    self.size = size
    self.field = field
    self.value = value

  def to_dict(self):
    '''
      :returns:
    '''
    return {
      'successful': self.successful,
      'error_msg': self.error_msg,
      'blob_key': str(self.blob_key),
      'name': self.name,
      'type': self.type,
      'size': self.size,
      # these two are commented out so the class is easily json serializable..
      # 'field': self.field,
      # 'value': self.value,
    }

def upload_blobs(validations=None):
  '''Method decorator for writing posted files to the `blobstore` using the
  App Engine files api. Passes an argument to the method with a list of
  `FieldResult` with `BlobKey`, name, type, size for each posted input file.

    :param validations:
  '''
  def wrapper(func):
    @wraps(func)
    def decorated(*args, **kw):
      return func(blobs=save_blobs(
        fields=_upload_fields(), validations=validations), *args, **kw)
    return decorated
  return wrapper

def save_blobs(fields, validations=None):
  '''Returns a list of `FieldResult` with BlobKey, name, type, size for
  each posted file.

    :param fields:
    :param validations:

    :returns:
  '''
  results = FieldResultSet()
  i = 0
  # validate_file_type=False,
  # validate_min_file_size=False,
  # validate_max_file_size=False
  for name, field in fields:
    value = field.stream.read()
    result = FieldResult(
      name=re.sub(r'^.*\\', '', field.filename),
      type=field.content_type,
      size=len(value),
      field=field,
      value=value)
    if validations:
      for fn in validations:
        if not fn(result):
          result.successful = False
          result.error_msg = MSG_INVALID_FILE_POSTED
          logging.warn('Error in file upload: %s', result.error_msg)
        else:
          result.blob_key = write_to_blobstore(
            result.value, mime_type=result.type, name=result.name)
          if result.blob_key:
            result.successful = True
          else:
            result.successful = False
          results.append(result)
    else:
      result.blob_key = write_to_blobstore(
        result.value, mime_type=result.type, name=result.name)
      if result.blob_key:
        result.successful = True
      else:
        result.successful = False
      results.append(result)
    i += 1
  return results

def _upload_fields():
  '''
    :returns: Returns a list of tuples with name of the file & stream as value.
  '''
  result = []
  for key, value in request.files.iteritems():
    if type(value) is not unicode:
      result.append((key, value))
  return result

def get_field_size(field):
  '''
    :param field:

    :returns:
  '''
  try:
    field.seek(0, 2) # Seek to the end of the file
    size = field.tell() # Get the position of EOF
    field.seek(0) # Reset the file position to the beginning
    return size
  except:
    return 0

def validate(field,
    accept_file_types=UPLOAD_ACCEPT_FILE_TYPES,
    min_file_size=UPLOAD_MIN_FILE_SIZE,
    max_file_size=UPLOAD_MAX_FILE_SIZE):
  '''Validates a file input based on size & type. If validation fails, adds
  an error property to the field arg dict.

    :param field:
    :param accept_file_types:
    :param min_file_size:
    :param max_file_size:

    :returns:
  '''
  if field.size < UPLOAD_MIN_FILE_SIZE:
    field.error_msg = 'min_file_size'
  elif field.size > UPLOAD_MAX_FILE_SIZE:
    field.error_msg = 'max_file_size'
  # only allow images to be posted to this handler
  elif not accept_file_types.match(field.type):
    field.error_msg = 'accept_file_types'
  else:
    return True
  return False

def write_to_blobstore(data, mime_type, name=None):
  '''Writes a file to the App Engine blobstore and returns an instance of a
  BlobKey if successful.

    :param data:
    :param mime_type:
    :param name:

    :returns:
  '''
  # if not name:
  #   name = ''.join(random.choice(string.letters)
  #     for x in range(DEFAULT_NAME_LEN))
  try:
    blob = files.blobstore.create(
      mime_type=mime_type,
      _blobinfo_uploaded_filename=name)
    with files.open(blob, 'a', exclusive_lock=True) as f:
      f.write(data)
    files.finalize(blob)
    result = files.blobstore.get_blob_key(blob)
    # issue with the local development SDK. we can only write to the blobstore
    # so fast, so set a retry_count and delay the execution thread between
    # each attempt..
    for i in range(1, WRITE_MAX_RETRIES):
      if result:
        break
      else:
        logging.debug('blob still None.. will retry to write to blobstore..')
        time.sleep(WRITE_SLEEP_SECONDS)
        result = files.blobstore.get_blob_key(blob)
      logging.debug('File written to blobstore: key: "%s"', result)
    return result
  except:
    import traceback
    logging.exception('Exception writing to the blobstore: %s',
      traceback.format_exc())
